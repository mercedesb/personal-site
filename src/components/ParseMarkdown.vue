<template>
  <div class="Markdown" v-html="html"></div>
</template>

<script>
import markdownItAnchor from "markdown-it-anchor";

var md = require("markdown-it")();

md.use(require("markdown-it-attrs"));
md.use(markdownItAnchor);

export default {
  props: {
    source: {
      type: String,
      required: true,
    },
    collapsible: {
      type: Boolean,
      default: false,
    },
    collapsibleTag: String,
    collapsedByDefault: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      parentSelector: "Collapsible",
      childSelector: "CollapsibleChild",
      modifierCollapsed: "--collapsed",
      modifierExpanded: "--expanded",
    };
  },
  computed: {
    querySelector: function() {
      return `.Markdown ${this.collapsibleTag}`;
    },
    html: function() {
      if (this.source) {
        const html = md.render(this.source);
        return html;
      }
      return "";
    },
  },
  mounted() {
    if (this.collapsible) {
      const collapsible = this.$el.querySelectorAll(this.querySelector);

      collapsible.forEach((c) => {
        c.addEventListener("click", this.handleCollapsible);

        if (this.collapsedByDefault) {
          this.handleCollapsible({ target: c }, this.modifierCollapsed);
        } else {
          this.handleCollapsible({ target: c }, this.modifierExpanded);
        }
      });
    }
  },
  methods: {
    handleCollapsible(event, modifier) {
      let el = event.target;

      this.toggleClass(el, this.parentSelector, modifier);

      let sibling = el.nextElementSibling;
      while (sibling) {
        if (sibling.matches(this.querySelector)) {
          break;
        }
        this.toggleClass(sibling, this.childSelector, modifier);
        sibling = sibling.nextElementSibling;
      }
    },
    toggleClass(el, selector, modifier) {
      if (!!selector && !!modifier) {
        el.className = `${el.className} ${selector} ${selector}${modifier}`;
      } else if (el.className.indexOf(this.modifierCollapsed) !== -1) {
        el.className = el.className.replace(
          this.modifierCollapsed,
          this.modifierExpanded
        );
      } else if (el.className.indexOf(this.modifierExpanded) !== -1) {
        el.className = el.className.replace(
          this.modifierExpanded,
          this.modifierCollapsed
        );
      }
    },
  },
};
</script>
<style lang="scss">
@import "../assets/styles/variables.scss";

$icon-dimension: 24px;

.Collapsible {
  cursor: pointer;

  &:before {
    content: "";
    display: inline-block;
    width: $icon-dimension;
    height: $icon-dimension;
    background-size: $icon-dimension $icon-dimension;
    margin-right: $base-spacing;
  }

  &--collapsed {
    &:before {
      background-image: url(../assets/ExpandIcon.svg);
    }
  }

  &--expanded {
    &:before {
      background-image: url(../assets/CollapseIcon.svg);
    }
  }

  &Child--collapsed {
    display: none;
  }

  &Child--expanded {
    display: block;
  }
}
</style>
