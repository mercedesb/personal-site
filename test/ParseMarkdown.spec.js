import ParseMarkdown from '@/components/ParseMarkdown.vue'
import Setup from './Setup'

describe('ParseMarkdown', () => {
  let component

  Setup.configure()

  const initialProps = {
    source: '### here is the source md\ntrying it out',
    collapsible: true,
    collapsibleTag: 'h3',
    collapsedByDefault: true
  }

  const shallow = (propsData, mocks) => Setup.shallow(ParseMarkdown, { 
    propsData: {
      ...initialProps,
      ...propsData
    }
  })
  
  describe('Snapshots', () => {
    it('matches snapshot', () => {
      component = shallow()
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('Properties', () => {
    it('has a source property', () => {
      component = shallow()
      expect(component.props().source).toEqual(initialProps.source)
    })

    it('has a collapsible property', () => {
      component = shallow()
      expect(component.props().collapsible).toEqual(true)
    })

    it('has a collapsibleTag property', () => {
      component = shallow()
      expect(component.props().collapsibleTag).toEqual('h3')
    })

    it('has a collapsedByDefault property', () => {
      component = shallow()
      expect(component.props().collapsedByDefault).toEqual(true)
    })
  })

  describe('Computed', () => {
    describe('querySelector', () => {
      it('returns Markdown and the collapsibleTag', () => {
        component = shallow()
        expect(component.vm.querySelector).toEqual('.Markdown h3')
      })
    })

    describe('html', () => {
      // TODO
      // describe('with source', () => {
      //   it('returns html markup', () => {

      //   })
      // })

      describe('without source', () => {
        it('returns the empty string', () => {
          component = shallow({source: undefined})
          expect(component.vm.html).toEqual('')
        })
      })
    })
  })

// TODO
  // describe('Lifecyle', () => {
  //   describe('mounted', () => {
  //     it('adds event listeners', () => {

  //     })

  //     describe('when collapsed by default', () => {
  //       it('calls handleCollapsible with collapsed modifier', () => {
  //         component = shallow()
  //       })
  //     })

  //     describe('when expanded by default', () => {
  //       it('calls handleCollapsible with expanded modifier', () => {
  //         component = shallow({collapsedByDefault: false})
  //       })
  //     })
  //   })
  // })

  // describe('Methods', () => {
  //   describe('handleCollapsible', () => {

  //   })

  //   describe('toggleClass', () => {

  //   })
  // })
})