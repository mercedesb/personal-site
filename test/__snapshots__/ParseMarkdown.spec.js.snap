// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParseMarkdown Snapshots matches snapshot 1`] = `
<div
  class="Markdown"
>
  <h3
    class=" Collapsible Collapsible--collapsed"
  >
    here is the source md
  </h3>
  

  <p
    class=" CollapsibleChild CollapsibleChild--collapsed"
  >
    trying it out
  </p>
  

</div>
`;
